docs(dagstore): add lotus rollback guidance to version prior to the dagstore (#1016)
filecoin-project/filecoin-docs: docs/mine/lotus/dagstore.md

## Conceptual overview

::: tip
The dagstore is targeted to be introduced in Lotus v1.11.2. We will update the docs once it’s confirmed.
:::

The dagstore is a sharded store to hold large IPLD graphs efficiently, packaged as location-transparent attachable CAR files, with mechanical sympathy resulting in zero-copy access in ideal situations.

> The dagstore is a component of the market's subsystem in `lotus-miner`. It replaces the former Badger staging blockstore. It is designed to provide high efficiency and throughput, and minimize resource utilization during deal-making operations.

Reviewer: It's actually genuinely called "markets subsystem":

```suggestion
The dagstore is a component of the _markets_ subsystem in `lotus-miner`. It replaces the former Badger staging blockstore. It is designed to provide high efficiency and throughput, and minimize resource utilization during deal-making operations.
```

- **Shard state:** the state that a dagstore shard is in. It can be one of the  following:
  - `ShardStateNew`: indicates that a shard has been registered, but has not yet been initialized.
  - `ShardStateInitializing`: indicates that the shard is being initialized.

>     - `ShardStateAvailable`: indicates that the shard has been initialized and is capable of serving retrievals. However, there are no active shard readers in this instance.

Reviewer: The previous meaning was correct, but just in case I'll use a synonym:

```suggestion
    - `ShardStateAvailable`: indicates that the shard has been initialized and is capable of serving retrievals. However, there are no active shard readers in this moment.
```

  - `ShardStateServing`: indicates the shard has active readers, and thus is currently serving at least one retrieval.
    - `ShardStateErrored`: indicates that an unexpected error was encountered during a shard operation (e.g. initialization or acquisition), and therefore the shard needs to be recovered to be fully operational.
    - `ShardStateRecovering`: indicates that the shard is attempting to recover from an errored state.
- **Shard initialization:** the act of fetching the shard's data, indexing it and storing the index under the `dagstore/index` directory.
- **Shard acquisition:** the act of fetching the shard's data, joining it with an index, and building an accessor for shard data to serve retrievals.

> - **Index:** a mapping of CID to CAR file offset, associated with a shard.

Reviewer: If we use non-technical notation, then let's stress.

```suggestion
- **Index:** a mapping of _CID to CAR_ file offset, associated with a shard.
```

## First-time migration

> When you first start your lotus-miner or the market nodes if you’ve split it out using the lotus that introduced dagstore(we will add the exact version later), a migration process will register all shards in **lazy initialization** mode. As deals come in, shards are fetched and initialized just in time to serve the retrieval.

Reviewer: ```suggestion
When you first start your lotus-miner or the market nodes if you’ve split it out using the lotus that introduced dagstore (we will add the exact version later), a migration process will register all shards in **lazy initialization** mode. As deals come in, shards are fetched and initialized just in time to serve the retrieval.
```

- By default, only unsealed pieces will be indexed to avoid forcing unsealing jobs. To index also sealed pieces, use the `--include-sealed` flag.

::: tip
In our test environments, we found the migration to process 400-500 shards/deals per second.
:::

## Configuration

> The DAG store can be configured through the `config.toml` file of the node that runs the `market's` subsystem. Refer to the `[DAGStore]` section. Lotus ships with sane defaults:

Reviewer: ```suggestion
The DAG store can be configured through the `config.toml` file of the node that runs the _markets_ subsystem. Refer to the `[DAGStore]` section. Lotus ships with sane defaults:
```

1. If possible, configure your markets node `storage.json` to point to storage paths that are shared with your mining/sealing/storage node. That will avoid unnecessary network transfer.
2. Do plan and execute [bulk initialization](#forcing-bulk-initialization) ASAP.

## Troubleshooting

### Lotus Version Rollback

> If you are downgrading from Lotus >=1.11.2-dev to 1.11.1 or below, you will need to ensure outstanding storage deals have reached the `StorageDealAwaitingPrecommit` state, before downgrading. If downgrading is performed before outstanding storage deals reach this state, the market's process will panic. If you are experiencing this panic, you will need to restore to your previous version and all the storage deals to reach `StorageDealAwaitingPrecommit` before attempting to downgrade again.

Reviewer: ```suggestion

If you are downgrading from Lotus 1.11.2-dev (or above) to 1.11.1 (or below), you will need to ensure outstanding storage deals have reached the `StorageDealAwaitingPrecommit` state, before downgrading. That's the stage at which the _markets_ subsystem has handed off the deal to the _sealing_ subsystem.

If downgrading is performed before outstanding storage deals reach this state, the _markets_ process will panic.

If you are experiencing this panic, you will need to restore to your previous version and all the storage deals to reach `StorageDealAwaitingPrecommit` before attempting to downgrade again.
```
